src/xrpld/app/tx/detail/VaultCreate.cpp

```diff
@@ -0,0 +1,198 @@
+  Copyright (c) 2023 Ripple Labs Inc.
+
+  Permission to use, copy, modify, and/or distribute this software for any
+  purpose  with  or without fee is hereby granted, provided that the above
+  copyright notice and this permission notice appear in all copies.
+
+  THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+  WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+  MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+  ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+  WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+  ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <xrpld/app/tx/detail/MPTokenIssuanceCreate.h>
+#include <xrpld/app/tx/detail/VaultCreate.h>
+#include <xrpld/ledger/View.h>
+#include <xrpl/protocol/Feature.h>
+#include <xrpl/protocol/Indexes.h>
+#include <xrpl/protocol/MPTIssue.h>
+#include <xrpl/protocol/Protocol.h>
+#include <xrpl/protocol/STNumber.h>
+#include <xrpl/protocol/TER.h>
+#include <xrpl/protocol/TxFlags.h>
+
+namespace ripple {
+
+NotTEC
+VaultCreate::preflight(PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled(featureSingleAssetVault))
+        return temDISABLED;
+
+    if (ctx.tx.isFieldPresent(sfDomainID) &&
+        !ctx.rules.enabled(featurePermissionedDomains))
+        return temDISABLED;
+
+    if (auto const ter = preflight1(ctx))
+        return ter;
+
+    if (ctx.tx.getFlags() & tfVaultCreateMask)
+        return temINVALID_FLAG;
+
+    if (auto const data = ctx.tx[~sfData])
+    {
+        if (data->empty() || data->length() > maxVaultDataLength)
+            return temMALFORMED;
+    }
+
+    if (auto const data = ctx.tx[~sfWithdrawalPolicy])
+    {
+        // Enforce valid withdrawal policy
+        if (*data != vaultStrategyFirstComeFirstServe)
+            return temMALFORMED;
+    }
+
+    if (auto const domain = ctx.tx[~sfDomainID])
+    {
+        if (*domain == beast::zero)
+            return temMALFORMED;
+        else if ((ctx.tx.getFlags() & tfVaultPrivate) == 0)
+            return temMALFORMED;  // DomainID only allowed on private vaults
+    }
+
+    if (auto const metadata = ctx.tx[~sfMPTokenMetadata])
+    {
+        if (metadata->length() == 0 ||
+            metadata->length() > maxMPTokenMetadataLength)
+            return temMALFORMED;
+    }
+
+    return preflight2(ctx);
+}
+
+XRPAmount
+VaultCreate::calculateBaseFee(ReadView const& view, STTx const& tx)
+{
+    // One reserve increment is typically much greater than one base fee.
+    return view.fees().increment;
+}
+
+TER
+VaultCreate::preclaim(PreclaimContext const& ctx)
+{
+    auto asset = ctx.tx[sfAsset];
+    auto account = ctx.tx[sfAccount];
+
+    if (asset.holds<MPTIssue>())
```

Comment:
I really like what Greg and Shawn did with the visit where they separated the checks for IOU vs MPT. Makes it very easy to review. Please think about implementing it. See my comment about EscrowToken for implementation details.

Reply:
which comment specifically ?